### PR TITLE
Outsource magic error messages

### PR DESCRIPTION
**Pull request by repasics**
Modify tools/gen-magic-strings.py to generate error messages.

JerryScript-DCO-1.0-Signed-off-by: Csaba Repasi repasics@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-errors.h [hunk 1]
```diff
@@ -13,82 +13,23 @@
...
+/** @cond doxygen_suppress */
+#define ECMA_ERROR_DEF(id, ascii_zt_string) id,
+#include "ecma-err-msg.inc.h"
+#undef ECMA_ERROR_DEF
+  /** @endcond */
+  ECMA_IS_VALID_CONSTRUCTOR
+} ecma_error_id_t;
+
+const lit_utf8_byte_t* get_ecma_error_utf8 (uint32_t id);
+lit_utf8_size_t get_ecma_error_size (uint32_t id);
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.h:33]: Should this be `ecma_get_...` ? (afaik we don't have any  `get_` prefixed method)

File: jerry-core/ecma/base/ecma-errors.c [hunk 2]
```diff
@@ -15,202 +15,53 @@
...
+{
+  JERRY_ASSERT (id != ECMA_IS_VALID_CONSTRUCTOR);
+#if JERRY_ERROR_MESSAGES
+  static const lit_utf8_byte_t *const ecma_error[] JERRY_ATTR_CONST_DATA = {
+/** @cond doxygen_suppress */
+#define ECMA_ERROR_DEF(id, utf8_string) (const lit_utf8_byte_t *) utf8_string,
+#include "ecma-err-msg.inc.h"
+#undef ECMA_ERROR_DEF
+    /** @endcond */
+  };
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.c:36]: * Is there a reason to have this array inside the method?
* Is there any reason not to only include the error messages inc.h once? Like for example just before the methods we could create a struct which have both the strings and size:
```c
#if JERRY_ERROR_MESSAGES

static const struct {
  list_utf8_byte_t *const text;
  uint32_t size;
} ecma_error_messages[] JERRY_ATTR_CONST_DATA = {
/** @cond doxygen_suppress */
#define ECMA_ERROR_DEF(id, utf8_string) \
  { (const lit_utf8_byte_t *) utf8_string, sizeof (utf8_string) - 1 },

#include "ecma-err-msg.inc.h"

#undef ECMA_ERROR_DEF
/** @endcond */
};
#endif

```

File: jerry-core/ecma/base/ecma-errors.h [hunk 3]
```diff
@@ -13,82 +13,23 @@
...
+#include "lit-globals.h"
+
+typedef enum
+{
+  ECMA_ERR_EMPTY,
+/** @cond doxygen_suppress */
+#define ECMA_ERROR_DEF(id, ascii_zt_string) id,
+#include "ecma-err-msg.inc.h"
+#undef ECMA_ERROR_DEF
+  /** @endcond */
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.h:28]: This indentation seems strange

File: jerry-core/ecma/base/ecma-errors.h [hunk 4]
```diff
@@ -13,82 +13,23 @@
...
-extern const char* const ecma_error_tagged_template_literals;
-#endif /* !ECMA_ERRORS_H */
+#ifndef ECMA_ERROR_H
+#define ECMA_ERROR_H
+
+#include "lit-globals.h"
+
+typedef enum
+{
+  ECMA_ERR_EMPTY,
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.h:23]: AFAIK: With this here, the array(s) in the errors.cpp file fill be all off by one.

File: jerry-core/ecma/base/ecma-errors.c [hunk 5]
```diff
@@ -15,202 +15,53 @@
...
+
+  return ecma_error[id];
+#else /* !JERRY_ERROR_MESSAGES */
+  return (const lit_utf8_byte_t *) "";
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-errors.c:40]: NULL should be used, shouldn't it?

**Review by zherczeg — CHANGES_REQUESTED** [review]
Very good patch. Are these strings available as magic strings? Since they are part of the binary anyway, there should be no memory allocation when these strings are used in JS code.

**Comment by repasics** [comment]
PR was refreshed: Generate error messages with magic strings using the same ini file.

File: tools/gen-magic-strings.py [hunk 6]
```diff
@@ -161,10 +167,10 @@ def calculate_magic_string_guards(defs, uses, debug=False):
...
 
     for str_ref, str_value in defs:
-        if str_ref not in uses:
-            print('warning: unused magic string {str_ref}'
-                  .format(str_ref=str_ref))
-            continue
+        # if str_ref not in uses:
+        #     print('warning: unused magic string {str_ref}'
+        #           .format(str_ref=str_ref))
+        #     continue
```
Comment (galpeter) [inline tools/gen-magic-strings.py:173]: Is it ok to have this commented out? Shouldn't we keep this?

Comment (repasics) [reply tools/gen-magic-strings.py:173]: Yes we should keep it. I have left it commented accidentally.

File: tests/unit-core/test-to-length.c [hunk 7]
```diff
@@ -48,7 +50,9 @@ main (void)
   TEST_ASSERT (num == 123);
 
   /* 1, 3 */
-  ecma_value_t error_throw = ecma_raise_type_error (ECMA_ERR_MSG ("I'm an error"));
+  char msg[] = "I'm an error";
+
+  ecma_value_t error_throw = ecma_raise_standard_error (JERRY_ERROR_TYPE, (const lit_utf8_byte_t *) msg, sizeof (msg));
```
Comment (rerobika) [inline tests/unit-core/test-to-length.c:55]: msg -> message_p
sizeof (msg) -> sizeof (message_p) - 1

File: tests/unit-core/test-to-integer.c [hunk 8]
```diff
@@ -48,7 +49,9 @@ main (void)
   TEST_ASSERT (num == 123);
 
   /* 2 */
-  ecma_value_t error = ecma_raise_type_error (ECMA_ERR_MSG ("I am a neat little error message"));
+  char msg[] = "I am a neat little error message";
```
Comment (rerobika) [inline tests/unit-core/test-to-integer.c:52]: same as below

File: jerry-core/ecma/base/ecma-errors.c [hunk 9]
```diff
@@ -16,201 +16,57 @@
...
-
-#endif /* JERRY_ESNEXT */
-
+ * Struct to store ecma error message with its size.
+ */
+typedef struct
+{
+  lit_utf8_byte_t *text; /* Text of ecma error message. */
+  lit_utf8_size_t size; /* Size of ecma error message. */
+} ecma_error_message;
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-errors.c:26]: add `_t` suffix

File: jerry-core/ecma/base/ecma-errors.c [hunk 10]
```diff
@@ -16,201 +16,57 @@
...
-                                                          " initialized before reading their value");
-
-#endif /* JERRY_ESNEXT */
-
+ * Struct to store ecma error message with its size.
+ */
+typedef struct
+{
+  lit_utf8_byte_t *text; /* Text of ecma error message. */
+  lit_utf8_size_t size; /* Size of ecma error message. */
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-errors.c:25]: uint8_t

File: jerry-core/ecma/base/ecma-errors.c [hunk 11]
```diff
@@ -16,201 +16,57 @@
...
+} ecma_error_message;
+
+/* Error message texts with size. */
+static ecma_error_message ecma_error_messages[] JERRY_ATTR_CONST_DATA = {
+  { (lit_utf8_byte_t *) "", 0 }, /* ECMA_ERR_EMPTY */
+/** @cond doxygen_suppress */
+#define ECMA_ERROR_DEF(id, utf8_string) { (lit_utf8_byte_t *) utf8_string, sizeof (utf8_string) - 1 },
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-errors.c:32]: add 3rd size argument

**Review by rerobika — CHANGES_REQUESTED** [review]

**Review by rerobika — APPROVED** [review]

File: jerry-core/ecma/base/ecma-err-msg.inc.h [hunk 12]
```diff
@@ -0,0 +1,856 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-err-msg.inc.h:1]: The file name could be `ecma-error-messages.inc.h` ( and the related ini file also). I don't see any reason why not to type it out.

**Comment by repasics** [comment]
The PR is updated.

File: jerry-core/ecma/base/ecma-errors.h [hunk 13]
```diff
@@ -13,82 +13,27 @@
...
-#endif /* JERRY_ESNEXT */
-
+#define ECMA_ERROR_DEF(id, ascii_zt_string) id,
+#else /* !JERRY_ERROR_MESSAGES */
+#define ECMA_ERROR_DEF(id, ascii_zt_string) id = ECMA_ERR_EMPTY,
 #endif /* JERRY_ERROR_MESSAGES */
+#include "ecma-error-messages.inc.h"
+#undef ECMA_ERROR_DEF
+  /** @endcond */
+  ECMA_IS_VALID_CONSTRUCTOR
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.h:33]: Is it ok tho have this here? I don't see a relevant entry in the .c file.

File: jerry-core/ecma/base/ecma-errors.h [hunk 14]
```diff
@@ -13,82 +13,27 @@
  * limitations under the License.
  */
 
-#include "config.h"
+#ifndef ECMA_ERROR_H
+#define ECMA_ERROR_H
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.h:17]: Shouldn't this be "...ERRORS..." as the file name contains "errors"

File: jerry-core/ecma/base/ecma-globals.h [hunk 15]
```diff
@@ -2488,4 +2488,4 @@ typedef struct
  * @}
  */
 
-#endif /* !ECMA_GLOBALS_H */
+#endif /* ECMA_GLOBALS_H */
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-globals.h:2491]: Is this correct to remove the `!` ?

**Review by galpeter — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/base/ecma-errors.h (hunk 13 above)
Comment (repasics) [reply jerry-core/ecma/base/ecma-errors.h:33]: It is used in ecma-function-object.c

File: jerry-core/ecma/base/ecma-errors.h [hunk 16]
```diff
@@ -13,82 +13,27 @@
...
+#if JERRY_ERROR_MESSAGES
+#define ECMA_ERROR_DEF(id, ascii_zt_string) id,
+#else /* !JERRY_ERROR_MESSAGES */
+#define ECMA_ERROR_DEF(id, ascii_zt_string) id = ECMA_ERR_EMPTY,
 #endif /* JERRY_ERROR_MESSAGES */
+#include "ecma-error-messages.inc.h"
+#undef ECMA_ERROR_DEF
+  /** @endcond */
+  ECMA_IS_VALID_CONSTRUCTOR
+} ecma_error_id_t;
```
Comment (dbatyai) [inline jerry-core/ecma/base/ecma-errors.h:34]: I don't really like the name for this, it does not identify the error, only the message. How about calling this `ecma_error_msg_t`

File: jerry-core/ecma/operations/ecma-exceptions.c [hunk 17]
```diff
@@ -414,9 +414,9 @@ ecma_raise_standard_error_with_format (jerry_error_t error_type, /**< error type
  *         Returned value must be freed with ecma_free_value
  */
 ecma_value_t
-ecma_raise_common_error (const char *msg_p) /**< error message */
+ecma_raise_common_error (ecma_error_id_t msg_p) /**< error message */
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-exceptions.c:417]: The argument is no longer a pointer, the _p suffix should be removed.

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 18]
```diff
@@ -155,7 +156,7 @@ ecma_op_is_callable (ecma_value_t value) /**< ecma value */
  * @return ECMA_IS_VALID_CONSTRUCTOR - if object is a valid for constructor call
  *         any other value - if object is not a valid constructor, the pointer contains the error message.
  */
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-function-object.c:158]: The "any other" value" comment is now invalid, no pointer is returned here. Please update.

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 19]
```diff
@@ -266,12 +267,12 @@ ecma_object_check_constructor (ecma_object_t *obj_p) /**< ecma object */
  * @return ECMA_IS_VALID_CONSTRUCTOR - if the input value is a constructor.
  *         any other value - if the input value is not a valid constructor, the pointer contains the error message.
  */
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-function-object.c:269]: The "any other" value" comment is now invalid, no pointer is returned here. Please update.

File: tools/gen-strings.py [hunk 20]
```diff
@@ -57,38 +62,45 @@ def read_magic_string_defs(debug=False):
...
-        for str_ref in re.findall('LIT_MAGIC_STRING_[a-zA-Z0-9_]+', line):
-            if str_ref in ['LIT_MAGIC_STRING_DEF',
-                           'LIT_MAGIC_STRING_FIRST_STRING_WITH_SIZE',
-                           'LIT_MAGIC_STRING_LENGTH_LIMIT',
-                           'LIT_MAGIC_STRING__COUNT']:
+        for str_ref in re.findall('%s_[a-zA-Z0-9_]+' % pattern, line):
+            if str_ref in ['%s_DEF' % pattern,
+                           '%s_FIRST_STRING_WITH_SIZE' % pattern,
+                           '%s_LENGTH_LIMIT' % pattern,
+                           '%s__COUNT' % pattern]:
```
Comment (galpeter) [inline tools/gen-strings.py:97]: This list should be constructed before the loop and there is should be just used.

File: jerry-core/ecma/base/ecma-errors.c [hunk 21]
```diff
@@ -16,201 +16,57 @@
...
+  return ecma_error_messages[id].text;
+#else /* !JERRY_ERROR_MESSAGES */
+  return NULL;
+#endif /* JERRY_ERROR_MESSAGES */
 
-/**
- * Error message, Cannot allocate memory for literals
- */
-const char* const ecma_error_cannot_allocate_memory_literals = "Cannot allocate memory for literals";
+} /* ecma_get_error_utf8 */
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-errors.c:55]: Was this newline always here?

**Review by galpeter — APPROVED** [review]
Minor change request, otherwise lgtm

File: jerry-core/ecma/base/ecma-errors.c (hunk 21 above)
Comment (repasics) [reply jerry-core/ecma/base/ecma-errors.c]: I removed it.

**Review by rerobika — APPROVED** [review]

**Review by zherczeg — APPROVED** [review]
LGTM

**Pull request merged**